### DISCUSSION TITLE
KDF master key length

### DISCUSSION THREAD

**Discussion by KevinRoebert** [jedisct1/libsodium#1209] [Q&A]
In the documentation for [libsodium < 1.0.12](https://doc.libsodium.org/key_derivation#key-derivation-with-libsodium-less-than-1.0.12) there exist an example to derive a subkey with the `crypto_generichash_blake2b_salt_personal` function. This example uses a **64 byte** master key.

The wrapper for KDF in [libsodium >= 1.0.12](https://doc.libsodium.org/key_derivation#key-derivation-with-libsodium-greater-than-1.0.12) uses only **32 bytes** for the master key: https://github.com/jedisct1/libsodium/blob/70c2796ae57979172ae882d60923d15b2aa25327/src/libsodium/include/sodium/crypto_kdf_blake2b.h#L29

Is this intended and safe? Why uses the old example 64 bytes and the wrapper only 32 bytes for the master key?

**Comment by KevinRoebert** [comment]
**A follow-up question**: is it safe to build a key hierarchy with this KDF feature? In other words, to generate several _second-level key-derivation keys_ from a _master key_ and from these then the final keys for the encryption as described in "_5. Key Hierarchy_" in [NIST SP 800-108r1](https://csrc.nist.gov/publications/detail/sp/800-108/rev-1/final)? Is there a maximum depth for such a hierarchy?

**Comment by KevinRoebert** [comment]
@jedisct1 Any information on this?

**Comment by jedisct1** [comment]
512 bits is the maximum key length supported bu the BLAKE2b function. The purpose of this example was to show that the master key and derived keys can have different sizes. But unless they are not uniformly sampled, 512 bit keys are useless, and 256 bits is more than enough.

The maximum depth should be the same as the cycle length. With 256 bit keys, it's over 2^127, which really means that there is no practical limits. With 128 bit keys, it's over 2^63.
